File: LeetCode/328. Odd Even Linked List.py
###  KNOWLEDGE IS POWER ###

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        
        even_head = head.next
        odd, even = head, head.next
        
        while odd.next and odd.next.next:
            odd.next = odd.next.next
            even.next = even.next.next
            odd, even = odd.next, even.next
        
        odd.next = even_head
        return head

File: LeetCode/328. Odd Even Linked List.cpp
/***  KNOWLEDGE IS POWER  ***/

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(!head || !head->next || !head->next->next) 
            return head;
        
        ListNode *odd = head ;
        ListNode *even = head->next ;
        ListNode *even_start = head->next ;
        
        while(odd->next && even->next)
        {
            odd->next = even->next ;
            even->next = odd->next->next ;
            odd = odd->next ;
            even = even->next ;
        }
        odd->next = even_start ;
        return head ; 
    }
};
